### PR TITLE
[FLINK-15782] [connectors/jdbc] JDBC sink DataStream API

### PR DESCRIPTION
**Pull request by rkhachatryan**
## What is the purpose of the change

1. expose existing sinks at the DataStream API level (for consistency with the planned exactly once; and to use on its own); existing API shouldn’t change (mostly table API)
2. refactor current code to allow re-use (in the planned exactly once JDBC sink); 
some issues are
- it operates on `Row` objects and (for upsert) 
- expects a tuple of `<boolean, Row>`
- generation of SQL doesn't make sense for POJO, because parameters must be set explicitly
- upsert emulation (`select` + `update` OR `isnert`) doesn't make sense for DataStream API

## Out of scope
- Table API
- existing implementation classes not required to implement the new API

## Brief change log
  - *extract connection management from `OutputFormat` to new `JdbcConnectionProvider`*
  - *replace `Row` with type parameter; move Table API specific code (like Tuple2<Boolean, T>) to the appropriate classes; as a result*:
   -- `GenericJdbcSinkFunction` (from `JDBCUpsertSinkFunction`)
   -- `JdbcBatchingOutputFormat` (from `JDBCUpsertOutputFormat`)
  - *add DataStream API (facade class)*
  - *add docs*
  - *minor changes: (change case `JDBC` -> `Jdbc`, parameter objects)*
  - *refactor tests*

## Verifying this change

This change is already covered by existing tests, such as *JDBCFullTest, JDBCOutputFormatTest, JdbcTableOutputFormatTest*.

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): no
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: yes
  - The serializers: no
  - The runtime per-record code paths (performance sensitive): yes
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Yarn/Mesos, ZooKeeper: no
  - The S3 file system connector: no

## Documentation

  - Does this pull request introduce a new feature? yes
  - If yes, how is the feature documented? docs

### CODE REVIEW

File: docs/redirects/jdbc.md [hunk 1]
```diff
@@ -0,0 +1,24 @@
+---
```
Comment (aljoscha) [inline docs/redirects/jdbc.md]: Why are these redirects added?

Comment (rkhachatryan) [reply docs/redirects/jdbc.md]: Just for uniformity with other pages.

Comment (aljoscha) [reply docs/redirects/jdbc.md]: The redirects are only there because previously the documentation structure was different. For new pages we don't need redirects.

File: flink-connectors/flink-jdbc/src/main/java/org/apache/flink/api/java/io/jdbc/JdbcStatementBuilder.java [hunk 2]
```diff
@@ -0,0 +1,46 @@
...
+ * @see JdbcBatchStatementExecutor
+ */
+@PublicEvolving
+public interface JdbcStatementBuilder<T> extends BiConsumerWithException<PreparedStatement, T, SQLException>, Serializable {
+
+	/**
+	 * Creates a {@link JdbcStatementBuilder} for {@link Row} using the provided SQL types array.
+	 * Uses {@link org.apache.flink.api.java.io.jdbc.JDBCUtils#setRecordToStatement}
+	 */
+	static JdbcStatementBuilder<Row> forRow(int[] types) {
```
Comment (aljoscha) [inline flink-connectors/flink-jdbc/src/main/java/org/apache/flink/api/java/io/jdbc/JdbcStatementBuilder.java]: I think in the long run we don't want to have `Row` in non-Table packages. We therefore shouldn't further expose it in our public APIs.

File: docs/dev/connectors/jdbc.md [hunk 3]
```diff
@@ -0,0 +1,97 @@
...
+under the License.
+-->
+
+* This will be replaced by the TOC
+{:toc}
+
+
+This connector provides a sink that writes data to a JDBC database.
+
+To use this it, add the following dependency to your project (along with your JDBC-driver):
```
Comment (kl0u) [inline docs/dev/connectors/jdbc.md]: Remove the "this".

File: flink-connectors/flink-jdbc/src/main/java/org/apache/flink/api/java/io/jdbc/JdbcBatchingOutputFormat.java [hunk 4]
```diff
@@ -0,0 +1,298 @@
...
+	private transient volatile Exception flushException;
+
+	JdbcBatchingOutputFormat(
+			JdbcConnectionProvider connectionProvider,
+			JdbcExecutionOptions executionOptions,
+			ExecutorCreator<JdbcExec> statementExecutorCreator,
+			RecordExtractor<In, JdbcIn> recordExtractor) {
+		super(connectionProvider);
+		this.executionOptions = executionOptions;
+		this.statementRunnerCreator = statementExecutorCreator;
```
Comment (kl0u) [inline flink-connectors/flink-jdbc/src/main/java/org/apache/flink/api/java/io/jdbc/JdbcBatchingOutputFormat.java]: Maybe uniformise the names, either use `runner` or `executor` for both the class (type) and the variable (instance). It is just easier to follow the code I think.

File: flink-connectors/flink-jdbc/src/main/java/org/apache/flink/api/java/io/jdbc/JdbcBatchingOutputFormat.java [hunk 5]
```diff
@@ -0,0 +1,298 @@
...
+
+	/**
+	 * Connects to the target database and initializes the prepared statement.
+	 *
+	 * @param taskNumber The number of the parallel instance.
+	 */
+	@Override
+	public void open(int taskNumber, int numTasks) throws IOException {
+		super.open(taskNumber, numTasks);
+		jdbcStatementExecutor = statementRunnerCreator.apply(getRuntimeContext());
```
Comment (kl0u) [inline flink-connectors/flink-jdbc/src/main/java/org/apache/flink/api/java/io/jdbc/JdbcBatchingOutputFormat.java]: Maybe introduce a method like: 

```
private JdbcExec initStatementExecutor() throws IOException {
		final JdbcExec exec = statementRunnerCreator.apply(getRuntimeContext());
		try {
			jdbcStatementExecutor.open(connection);
		} catch (SQLException e) {
			throw new IOException("unable to open JDBC writer", e);
		}
		return exec;
	}
```

To offload a bit of the code in the `open()`. What do you think?

File: flink-connectors/flink-jdbc/src/main/java/org/apache/flink/api/java/io/jdbc/JdbcBatchingOutputFormat.java [hunk 6]
```diff
@@ -0,0 +1,298 @@
...
+		try {
+			jdbcStatementExecutor.open(connection);
+		} catch (SQLException e) {
+			throw new IOException("unable to open JDBC writer", e);
+		}
+		if (executionOptions.getBatchIntervalMs() != 0 && executionOptions.getBatchSize() != 1) {
+			this.scheduler = Executors.newScheduledThreadPool(1, new ExecutorThreadFactory("jdbc-upsert-output-format"));
+			this.scheduledFuture = this.scheduler.scheduleWithFixedDelay(() -> {
+				synchronized (JdbcBatchingOutputFormat.this) {
+					if (!closed) {
```
Comment (kl0u) [inline flink-connectors/flink-jdbc/src/main/java/org/apache/flink/api/java/io/jdbc/JdbcBatchingOutputFormat.java]: Are we sure about the concurrency here? Here `closed` is guarded by a lock but all the remaining usages are not. Here we are simply reading the value which is `volatile`, but in the `close()` we have a `if (!closed) { closed = true;  ... ` which is not atomic and then we `cancel` the `scheduledFuture` executing this code.

File: flink-connectors/flink-jdbc/src/main/java/org/apache/flink/api/java/io/jdbc/JdbcBatchingOutputFormat.java [hunk 7]
```diff
@@ -0,0 +1,298 @@
...
+		jdbcStatementExecutor.process(jdbcRecordExtractor.apply(record));
+	}
+
+	@Override
+	public synchronized void flush() throws IOException {
+		checkFlushException();
+
+		for (int i = 1; i <= executionOptions.getMaxRetries(); i++) {
+			try {
+				attemptFlush();
```
Comment (kl0u) [inline flink-connectors/flink-jdbc/src/main/java/org/apache/flink/api/java/io/jdbc/JdbcBatchingOutputFormat.java]: Here shouldn't we restart the timer? It seems like we may be in the situation where we flushed this batch because we reached the max size, so for the new one we have to set a new timer (renew the `BatchIntervalMs`), right?

File: flink-connectors/flink-jdbc/src/main/java/org/apache/flink/api/java/io/jdbc/dialect/JDBCDialects.java [hunk 8]
```diff
@@ -152,7 +152,7 @@ public String quoteIdentifier(String identifier) {
 			return identifier;
 		}
 
-		@Override
+@Override
```
Comment (kl0u) [inline flink-connectors/flink-jdbc/src/main/java/org/apache/flink/api/java/io/jdbc/dialect/JDBCDialects.java]: Probably a typo?

File: flink-connectors/flink-jdbc/src/main/java/org/apache/flink/api/java/io/jdbc/executor/KeyedBatchStatementExecutor.java [hunk 9]
```diff
@@ -0,0 +1,79 @@
...
+
+import org.apache.flink.api.java.io.jdbc.JdbcStatementBuilder;
+
+import java.sql.Connection;
+import java.sql.PreparedStatement;
+import java.sql.SQLException;
+import java.util.HashSet;
+import java.util.Set;
+import java.util.function.Function;
+
```
Comment (kl0u) [inline flink-connectors/flink-jdbc/src/main/java/org/apache/flink/api/java/io/jdbc/executor/KeyedBatchStatementExecutor.java]: Here a javadoc for the types and the functionality would help. I noticed that none of the classes have one, so it may be not part of this PR, but it would be a helpful addition I think.

File: flink-connectors/flink-jdbc/src/main/java/org/apache/flink/api/java/io/jdbc/JdbcExecutionOptions.java [hunk 10]
```diff
@@ -0,0 +1,92 @@
...
+import org.apache.flink.annotation.PublicEvolving;
+import org.apache.flink.util.Preconditions;
+
+import java.io.Serializable;
+
+/**
+ * JDBC sink batch options.
+ */
+@PublicEvolving
+public class JdbcExecutionOptions implements Serializable {
```
Comment (wuchong) [inline flink-connectors/flink-jdbc/src/main/java/org/apache/flink/api/java/io/jdbc/JdbcExecutionOptions.java]: It seems that this option is only used for batch writing. `Execution` sounds like it also contains reading/query, because `PreparedStatement` also exposes a `executeQuery` method. 
Shall we rename it to `JdbcWritingOptions` or `JdbcSinkOptions`?

File: flink-connectors/flink-jdbc/src/main/java/org/apache/flink/api/java/io/jdbc/JdbcSink.java [hunk 11]
```diff
@@ -0,0 +1,63 @@
...
+
+import org.apache.flink.annotation.PublicEvolving;
+import org.apache.flink.api.java.io.jdbc.executor.JdbcBatchStatementExecutor;
+import org.apache.flink.streaming.api.functions.sink.SinkFunction;
+
+/**
+ * Facade to create JDBC {@link SinkFunction sinks}.
+ */
+@PublicEvolving
+public class JdbcSink {
```
Comment (wuchong) [inline flink-connectors/flink-jdbc/src/main/java/org/apache/flink/api/java/io/jdbc/JdbcSink.java]: Shall we expose a corresponding `JdbcSource`, or it will be introduced later?

File: flink-connectors/flink-jdbc/src/main/java/org/apache/flink/api/java/io/jdbc/JdbcBatchingOutputFormat.java (hunk 5 above)
Comment (rkhachatryan) [reply flink-connectors/flink-jdbc/src/main/java/org/apache/flink/api/java/io/jdbc/JdbcBatchingOutputFormat.java]: Yes, I think that's a good idea.

File: flink-connectors/flink-jdbc/src/main/java/org/apache/flink/api/java/io/jdbc/JdbcBatchingOutputFormat.java (hunk 6 above)
Comment (rkhachatryan) [reply flink-connectors/flink-jdbc/src/main/java/org/apache/flink/api/java/io/jdbc/JdbcBatchingOutputFormat.java]: `close()` is also `synchronized` so these sections can't execute simultaneously.

`volatile` seems unnecessary here; I didn't remove this modifier because I tried to keep the internal logic untouched where possible (performance-wise it doesn't matter since it's not in the hot path).

File: flink-connectors/flink-jdbc/src/main/java/org/apache/flink/api/java/io/jdbc/JdbcBatchingOutputFormat.java (hunk 7 above)
Comment (rkhachatryan) [reply flink-connectors/flink-jdbc/src/main/java/org/apache/flink/api/java/io/jdbc/JdbcBatchingOutputFormat.java]: Yes. I would also replace `Thread.sleep(1000 * i);` with just rescheduling (retaining blocking semantics of `write`).
But I'm not sure if such changes should be in this PR.
WDYT?

Comment (kl0u) [reply flink-connectors/flink-jdbc/src/main/java/org/apache/flink/api/java/io/jdbc/JdbcBatchingOutputFormat.java]: I agree that this seems to be more of a bug that deserves its own PR. Would you like to open a JIRA for this and assign it to you?

File: flink-connectors/flink-jdbc/src/main/java/org/apache/flink/api/java/io/jdbc/JdbcConnectionOptions.java [hunk 12]
```diff
@@ -0,0 +1,94 @@
...
+	}
+
+	public String getUsername() {
+		return username;
+	}
+
+	public String getPassword() {
+		return password;
+	}
+
```
[the remark this answers is not in the document]
Comment (rkhachatryan) [reply flink-connectors/flink-jdbc/src/main/java/org/apache/flink/api/java/io/jdbc/JdbcConnectionOptions.java]: This name clashes with `JDBCOptions#builder` which is part of current public API.
I didn't want to have similar methods named differently so I decided just to create builder with `new`.

[the remark this answers is not in the document]
Comment (rkhachatryan) [reply flink-connectors/flink-jdbc/src/main/java/org/apache/flink/api/java/io/jdbc/JdbcConnectionOptions.java]: This name clashes with `JDBCOptions#builder` which is part of current public API.
I didn't want to have similar methods named differently so I decided just to create builder with `new`.

File: flink-connectors/flink-jdbc/src/main/java/org/apache/flink/api/java/io/jdbc/executor/JdbcBatchStatementExecutor.java [hunk 13]
```diff
@@ -0,0 +1,102 @@
...
+ */
+@Internal
+public interface JdbcBatchStatementExecutor<T> {
+
+	/**
+	 * Open the writer by JDBC Connection. It can create Statement from Connection.
+	 */
+	void open(Connection connection) throws SQLException;
+
+	void process(T record) throws SQLException;
```
[the remark this answers is not in the document]
Comment (rkhachatryan) [reply flink-connectors/flink-jdbc/src/main/java/org/apache/flink/api/java/io/jdbc/executor/JdbcBatchStatementExecutor.java]: Yes, I think it's a good idea.

File: flink-connectors/flink-jdbc/src/test/java/org/apache/flink/api/java/io/jdbc/JDBCTestCheckpoint.java [hunk 14]
```diff
@@ -0,0 +1,35 @@
...
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.flink.api.java.io.jdbc;
+
+/**
+ * Holds id and indices of items in {@link JdbcTestFixture#TEST_DATA}.
+ */
+public class JDBCTestCheckpoint {
```
[the remark this answers is not in the document]
Comment (rkhachatryan) [reply flink-connectors/flink-jdbc/src/test/java/org/apache/flink/api/java/io/jdbc/JDBCTestCheckpoint.java]: Yes, this is a leftover after splitting the work.
Removed it, thanks.

File: flink-connectors/flink-jdbc/src/main/java/org/apache/flink/api/java/io/jdbc/JdbcSink.java (hunk 11 above)
Comment (rkhachatryan) [reply flink-connectors/flink-jdbc/src/main/java/org/apache/flink/api/java/io/jdbc/JdbcSink.java]: This PR only targets JDBC sinks.
For sources I think a similar refactoring would be required.

File: flink-connectors/flink-jdbc/src/main/java/org/apache/flink/api/java/io/jdbc/JdbcExecutionOptions.java (hunk 10 above)
Comment (rkhachatryan) [reply flink-connectors/flink-jdbc/src/main/java/org/apache/flink/api/java/io/jdbc/JdbcExecutionOptions.java]: `JdbcWritingOptions` doesn't seem to match `delete` queries.
And `JdbcSinkOptions` sounds too generic to me. I'd expect something like JDBC URL inside.

As for `executeQuery`, I don't see why `JdbcExecutionOptions` can't be used for it (batch could mean `limit` then).

File: flink-connectors/flink-jdbc/src/main/java/org/apache/flink/api/java/io/jdbc/executor/KeyedBatchStatementExecutor.java [hunk 15]
```diff
@@ -0,0 +1,79 @@
...
+	@Override
+	public void process(T record) {
+		batch.add(keyExtractor.apply(record));
+	}
+
+	@Override
+	public void executeBatch() throws SQLException {
+		if (!batch.isEmpty()) {
+			for (K entry : batch) {
+				parameterSetter.accept(st, entry);
```
[the remark this answers is not in the document]
Comment (rkhachatryan) [reply flink-connectors/flink-jdbc/src/main/java/org/apache/flink/api/java/io/jdbc/executor/KeyedBatchStatementExecutor.java]: You are right, thanks for pointing this out.
I'll change it to use `SimpleBatchStatementExecutor`.

File: flink-connectors/flink-jdbc/src/main/java/org/apache/flink/api/java/io/jdbc/JdbcExecutionOptions.java (hunk 10 above)
Comment (wuchong) [reply flink-connectors/flink-jdbc/src/main/java/org/apache/flink/api/java/io/jdbc/JdbcExecutionOptions.java]: I don't think batch could mean `limit`. From my understanding, `JdbcExecutionOptions` is just a tuning options which should affect the final result. And the `batchIntervalMs` doesn't make sense for readers. 
But if we dont' have a better name for it, I'm also fine with `JdbcExecutionOptions`.

File: flink-connectors/flink-jdbc/src/main/java/org/apache/flink/api/java/io/jdbc/JdbcBatchingOutputFormat.java [hunk 16]
```diff
@@ -46,15 +47,15 @@
 		}
 	}
 
-	interface ExecutorCreator<T extends JdbcBatchStatementExecutor<?>> extends Function<RuntimeContext, T>, Serializable {
+	interface StatementExecutorCreator<T extends JdbcBatchStatementExecutor<?>> extends Function<RuntimeContext, T>, Serializable {
```
Comment (wuchong) [inline flink-connectors/flink-jdbc/src/main/java/org/apache/flink/api/java/io/jdbc/JdbcBatchingOutputFormat.java]: What about `StatementExecutorFactory

File: flink-connectors/flink-jdbc/src/main/java/org/apache/flink/api/java/io/jdbc/JdbcExecutionOptions.java (hunk 10 above)
Comment (rkhachatryan) [reply flink-connectors/flink-jdbc/src/main/java/org/apache/flink/api/java/io/jdbc/JdbcExecutionOptions.java]: I agree, `JdbcExecutionOptions` isn't ideal, but other names aren't better.

File: flink-connectors/flink-jdbc/src/main/java/org/apache/flink/api/java/io/jdbc/JdbcBatchingOutputFormat.java (hunk 7 above)
Comment (rkhachatryan) [reply flink-connectors/flink-jdbc/src/main/java/org/apache/flink/api/java/io/jdbc/JdbcBatchingOutputFormat.java]: Yes, that's a good idea.
Created a ticket for that: https://issues.apache.org/jira/browse/FLINK-16328

**Comment by rkhachatryan** [comment]
Thanks a lot for reviewing @aljoscha  ,  @kl0u ,   @wuchong .
I've addressed the issues and currently waiting for another PR to pass CI (didn't push yet).

**Comment by rkhachatryan** [comment]
Please review the changes (I've added them as a separate commit).

**Comment by rkhachatryan** [comment]
I've rebased the PR onto the latest master.
@kl0u , @wuchong, could you please take a look?

File: flink-connectors/flink-jdbc/src/main/java/org/apache/flink/api/java/io/jdbc/executor/InsertOrUpdateJdbcExecutor.java [hunk 17]
```diff
@@ -0,0 +1,124 @@
...
+	private final Function<R, K> keyExtractor;
+	private final Function<R, V> valueMapper;
+
+	private transient PreparedStatement existStatement;
+	private transient PreparedStatement insertStatement;
+	private transient PreparedStatement updateStatement;
+	private transient Map<K, V> batch = new HashMap<>();
+
+	public InsertOrUpdateJdbcExecutor(
+		String existSQL,
```
Comment (kl0u) [inline flink-connectors/flink-jdbc/src/main/java/org/apache/flink/api/java/io/jdbc/executor/InsertOrUpdateJdbcExecutor.java]: What about one more tab on the arguments so that we can distinguish between the args and the body of the method?

File: flink-connectors/flink-jdbc/src/main/java/org/apache/flink/api/java/io/jdbc/executor/InsertOrUpdateJdbcExecutor.java [hunk 18]
```diff
@@ -0,0 +1,124 @@
...
+		String insertSQL,
+		String updateSQL,
+		JdbcStatementBuilder<K> existSetter,
+		JdbcStatementBuilder<V> insertSetter,
+		JdbcStatementBuilder<V> updateSetter,
+		Function<R, K> keyExtractor,
+		Function<R, V> valueExtractor) {
+		this.existSQL = existSQL;
```
Comment (kl0u) [inline flink-connectors/flink-jdbc/src/main/java/org/apache/flink/api/java/io/jdbc/executor/InsertOrUpdateJdbcExecutor.java]: I think it makes sense to add null checks here so that the contracts of the class are explicit.

File: flink-connectors/flink-jdbc/src/main/java/org/apache/flink/api/java/io/jdbc/GenericJdbcSinkFunction.java [hunk 19]
```diff
@@ -19,18 +19,18 @@
...
-class JDBCUpsertSinkFunction extends RichSinkFunction<Tuple2<Boolean, Row>> implements CheckpointedFunction {
-	private final JDBCUpsertOutputFormat outputFormat;
+import java.io.IOException;
 
-	JDBCUpsertSinkFunction(JDBCUpsertOutputFormat outputFormat) {
+class GenericJdbcSinkFunction<T> extends RichSinkFunction<T> implements CheckpointedFunction {
+	private final AbstractJdbcOutputFormat<T > outputFormat;
+
+	GenericJdbcSinkFunction(AbstractJdbcOutputFormat<T > outputFormat) {
 		this.outputFormat = outputFormat;
```
Comment (kl0u) [inline flink-connectors/flink-jdbc/src/main/java/org/apache/flink/api/java/io/jdbc/GenericJdbcSinkFunction.java]: Add a null check?

File: flink-connectors/flink-jdbc/src/main/java/org/apache/flink/api/java/io/jdbc/JdbcBatchingOutputFormat.java [hunk 20]
```diff
@@ -0,0 +1,316 @@
...
+	private transient ScheduledFuture<?> scheduledFuture;
+	private transient volatile Exception flushException;
+
+	JdbcBatchingOutputFormat(
+			JdbcConnectionProvider connectionProvider,
+			JdbcExecutionOptions executionOptions,
+			StatementExecutorFactory<JdbcExec> statementExecutorFactory,
+			RecordExtractor<In, JdbcIn> recordExtractor) {
+		super(connectionProvider);
+		this.executionOptions = executionOptions;
```
Comment (kl0u) [inline flink-connectors/flink-jdbc/src/main/java/org/apache/flink/api/java/io/jdbc/JdbcBatchingOutputFormat.java]: I would also add null checks here wherever needed.

File: flink-connectors/flink-jdbc/src/main/java/org/apache/flink/api/java/io/jdbc/JdbcBatchingOutputFormat.java [hunk 21]
```diff
@@ -0,0 +1,316 @@
...
+	interface StatementExecutorFactory<T extends JdbcBatchStatementExecutor<?>> extends Function<RuntimeContext, T>, Serializable {
+	}
+
+	private static final long serialVersionUID = 1L;
+
+	private static final Logger LOG = LoggerFactory.getLogger(JdbcBatchingOutputFormat.class);
+
+	private final JdbcExecutionOptions executionOptions;
+	private final StatementExecutorFactory<JdbcExec> statementExecutorFactory;
+	final RecordExtractor<In, JdbcIn> jdbcRecordExtractor;
```
Comment (kl0u) [inline flink-connectors/flink-jdbc/src/main/java/org/apache/flink/api/java/io/jdbc/JdbcBatchingOutputFormat.java]: What about making it private and adding a package private or protected getter. This is a question of personal taste though so feel free to ignore if you disagree :)

File: flink-connectors/flink-jdbc/src/main/java/org/apache/flink/api/java/io/jdbc/JdbcDmlOptions.java [hunk 22]
```diff
@@ -0,0 +1,127 @@
...
+	public String[] getFieldNames() {
+		return fieldNames;
+	}
+
+	public Optional<String[]> getKeyFields() {
+		return Optional.ofNullable(keyFields);
+	}
+
+	/**
+	 * JDBCUpsertOptionsBuilder.
```
Comment (kl0u) [inline flink-connectors/flink-jdbc/src/main/java/org/apache/flink/api/java/io/jdbc/JdbcDmlOptions.java]: Probably add sth like "A builder for..." I understand that this is kind of redundant but at least have sth more meaningful :P

File: flink-connectors/flink-jdbc/src/main/java/org/apache/flink/api/java/io/jdbc/JdbcDmlOptions.java [hunk 23]
```diff
@@ -0,0 +1,127 @@
...
+	 */
+	public static class JdbcDmlOptionsBuilder extends JDBCUpdateQueryOptionsBuilder<JdbcDmlOptionsBuilder> {
+		private String tableName;
+		private String[] fieldNames;
+		private String[] keyFields;
+		private JDBCDialect dialect;
+		private JDBCDialect customDialect;
```
Comment (kl0u) [inline flink-connectors/flink-jdbc/src/main/java/org/apache/flink/api/java/io/jdbc/JdbcDmlOptions.java]: This seems to not be used anywhere.

File: flink-connectors/flink-jdbc/src/main/java/org/apache/flink/api/java/io/jdbc/executor/JdbcBatchStatementExecutor.java [hunk 24]
```diff
@@ -16,29 +16,27 @@
...
 import java.sql.SQLException;
+import java.util.function.Function;
 
 /**
- * JDBCWriter used to execute statements (e.g. INSERT, UPSERT, DELETE).
+ * Executes the given JDBC statement in batch for the accumulated records.
  */
-public interface JDBCWriter extends Serializable {
+@Internal
+public interface JdbcBatchStatementExecutor<T> {
```
Comment (kl0u) [inline flink-connectors/flink-jdbc/src/main/java/org/apache/flink/api/java/io/jdbc/executor/JdbcBatchStatementExecutor.java]: Here all the methods are expected to only throw `SQLException` and then, in many cases, we wrap the exceptions in other types. Does it make sense to simply have these methods throw an `Exception`?

File: flink-connectors/flink-jdbc/src/main/java/org/apache/flink/api/java/io/jdbc/executor/InsertOrUpdateJdbcExecutor.java (hunk 17 above)
Comment (rkhachatryan) [reply flink-connectors/flink-jdbc/src/main/java/org/apache/flink/api/java/io/jdbc/executor/InsertOrUpdateJdbcExecutor.java]: How do I always mess with indentation 🤦‍♂️

File: flink-connectors/flink-jdbc/src/main/java/org/apache/flink/api/java/io/jdbc/JdbcBatchingOutputFormat.java (hunk 21 above)
Comment (rkhachatryan) [reply flink-connectors/flink-jdbc/src/main/java/org/apache/flink/api/java/io/jdbc/JdbcBatchingOutputFormat.java]: Hmm...I think we can have a different "template method" - `addToBatch` instead of `doWriteRecord`; so this field is accessed only in the base class and can be `private`:
```
@Override
void addToBatch(Tuple2<Boolean, Row> original, Row extracted) throws SQLException {
...
```

File: flink-connectors/flink-jdbc/src/main/java/org/apache/flink/api/java/io/jdbc/JdbcDmlOptions.java (hunk 22 above)
Comment (rkhachatryan) [reply flink-connectors/flink-jdbc/src/main/java/org/apache/flink/api/java/io/jdbc/JdbcDmlOptions.java]: OK :)

File: flink-connectors/flink-jdbc/src/main/java/org/apache/flink/api/java/io/jdbc/JdbcDmlOptions.java (hunk 23 above)
Comment (rkhachatryan) [reply flink-connectors/flink-jdbc/src/main/java/org/apache/flink/api/java/io/jdbc/JdbcDmlOptions.java]: Leftover from some previous iteration. 
Removed, thanks.

File: flink-connectors/flink-jdbc/src/main/java/org/apache/flink/api/java/io/jdbc/executor/JdbcBatchStatementExecutor.java (hunk 24 above)
Comment (rkhachatryan) [reply flink-connectors/flink-jdbc/src/main/java/org/apache/flink/api/java/io/jdbc/executor/JdbcBatchStatementExecutor.java]: On higher levels, it is wrapped because there the allowed exception is `IOException`, not `SQLException`. So we can only declare `throws IOException` without wrapping.

But IMO `SQLException` is better here because:
- caller can actually do something with it
- it has an appropriate abstraction level

Comment (rkhachatryan) [reply flink-connectors/flink-jdbc/src/main/java/org/apache/flink/api/java/io/jdbc/executor/JdbcBatchStatementExecutor.java]: After further looking at code, I noticed it was wrapped in one place into `RuntimeException` instead of `IOException` so I fixed wrapping.
